### PR TITLE
Here comes another monster: Text-rendering and more GL fixes.

### PR DESCRIPTION
**Pull request by Lord-Kamina**
### What does this PR do?

- Adds
  - Support for stroke-linecap, stroke-linejoin and stroke-miterlimit SVG attributes.
  - Revamps the code for recoloring text while singing, now enabling people to theme both fill and stroke and even changing opacity levels.
  - A handy new pair of functions in UnicodeUtil for converting text to uppercase and lowercase in an utf-8 friendly way. We should use those instead of the boost functions for other things now.
- Changes
  - Instead of iterating through every shader to update uniforms, we now use a Uniform Buffer Object to represent four different blocks of uniforms; we're using a bit more GPU memory this way (especially accounting for the idiotic offset aligning which means allocating 800 bytes for a 368 byte buffer) but we only need to update the values in one place and it will be reflected in all shaders.
  - Removed my hacky text-scaling code that added spaces to text in an attempt to account for the scaling during singing. Fixed by just making sure that the x_offset between syllables is enough to account for the size change.
  - Increased max mipmap level to 12 because... why not? I mean, it is possible to run the game in 4k now.
  - Use different texture min and mag filters for regular textures and text. For text, we'll never use bilinear (it looked awful with my recoloring code, plus from what I understand it's slower)
  - Changed the SVG lyric files a little bit; added a different stroke-color to the highlight and cleaned them up a little bit.
  - Removed some unused pieces of code from both the shaders and opengl_text.
- Fixes
  - (Hopefully) I finally got the idiotic index out of bounds error that would happen at the beginning of some songs (duets, usually). It was apparently the result of a race-condition between threads.
  - Fixes an ffmpeg deprecation warning (wonder if they'll ever get tired of trolling everybody else).
  - Fixes broken screenshots in high-dpi mode.
  - Some other minor changes, fixes and cleanup.

### Closes Issue(s)

#411 sorta.
This game would so very much benefit from some sprite batching; that should be a medium-term goal but I don't think I can do it alone.

### More

- [x] Added/updated documentation : Added some extra commenting here and there.

### CODE REVIEW

File: data/shaders/core.frag [hunk 1]
```diff
@@ -1,23 +1,37 @@
...
 } fragIn;
 
+out vec4 fragColor;
+
+layout (std140) uniform shaderMatrices {
+	mat4 projMatrix;
+	mat4 mvMatrix;
+	mat4 normalMatrix;
+	mat4 colorMatrix;
+} shaderMatrices;
```
Comment (nieknooijens) [inline data/shaders/core.frag]: schermafdruk van 2018-11-01 19-31-46
this is "already defined"

File: data/shaders/dancenote.vert [hunk 2]
```diff
@@ -1,19 +1,26 @@
...
+	mat4 normalMatrix;
+	mat4 colorMatrix;
+} shaderMatrices;
+
+layout (std140) uniform danceNote {
+	int noteType;
+	float hitAnim;
+	float clock;
+	float scale;
+	vec2 position;
```
Comment (nieknooijens) [inline data/shaders/dancenote.vert]: why do we use vec4's for nearly everything, but vec2's for dance? weren't dance notes supposed to be 3d models in the future?

File: game/color.cc [hunk 3]
```diff
@@ -52,24 +52,24 @@ Color::Color(std::string const& str) {
...
-	else if (name == "fuchsia") return Color(1.0, 0.06, 127/255.0);
-	else if (name == "yellow") return Color(1.0, 1.0, 0.0);
-	else if (name == "green") return Color(0.0, 1.0, 0.0);	
-	else if (name == "red") return Color(1, 0.0, 0.0);
-	else if (name == "blue") return Color(0.0, 43.75/255.0, 1.0);
-	else return Color(0.5, 0.5, 0.5);
+	if (name == "black") return Color(3.0/255.0, 3.0/255.0, 3.0/255.0, 1.0);
+	else if (name == "gray") return Color(24.0/255.0, 24.0/255.0, 24.0/255.0, 1.0);
+	else if (name == "white") return Color(1.0, 1.0, 1.0, 1.0);
+	else if (name == "aqua") return Color(0.0, 1.0, 1.0, 1.0);
```
Comment (nieknooijens) [inline game/color.cc]: so everything has alpha now? making it consistent with everything else. makes sense.

File: game/glshader.cc [hunk 4]
```diff
@@ -49,6 +51,40 @@ void Shader::dumpInfoLog(GLuint id) {
 	std::clog << logmsg << std::flush;
 }
 
+void Shader::bindUniformBlocks() {
```
Comment (nieknooijens) [inline game/glshader.cc]: We should add some comments here as I have no idea what an "uniform block" is.

File: data/shaders/core.frag (hunk 1 above)
Comment (Lord-Kamina) [reply data/shaders/core.frag]: What the blue fuck? Maybe I need to change the instance names between stages? 🤷🏼‍♂️

File: data/shaders/dancenote.vert (hunk 2 above)
Comment (Lord-Kamina) [reply data/shaders/dancenote.vert]: Don't know honestly.

File: game/color.cc (hunk 3 above)
Comment (Lord-Kamina) [reply game/color.cc]: Yeah. Less margin for error IMO.

File: game/glshader.cc (hunk 4 above)
Comment (Lord-Kamina) [reply game/glshader.cc]: Uniforms organized in "interface blocks" which are essentially structs as you can see in the shaders. Cons: use more GPU memory, pros: shared state between all shaders so you only modify it once.

**Comment by Lord-Kamina** [comment]
@nieknooijens Anything else you want to see changed?

**Comment by nieknooijens** [comment]
not really, looks ok now.

**Comment by Lord-Kamina** [comment]
I'll merge it then.

**Pull request merged**